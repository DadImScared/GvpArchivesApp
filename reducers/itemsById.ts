import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import * as actions from "../actions";
import { ADD_ITEMS } from "../actiontypes/itemsById";
import { IReducerState } from "./index";

export interface Item {
  item_id: string;
  title: string;
  link: string;
  category: string;
  language?: string;
  year?: string;
  issue?: string;
  directory?: string;
}

export interface ItemsByIdState {
  [key: string]: Item;
}

function itemsById(state: ItemsByIdState = {}, action: ActionType<typeof actions.itemsById>) {
  switch (action.type) {
    case ADD_ITEMS:
      return {
        ...state,
        ...action.payload.items
      };
    default:
      return state;
  }
}

export const getItemsById = createSelector(
    [(state: IReducerState) => state.itemsById],
    (itemsState) => itemsState
);

export default itemsById;
