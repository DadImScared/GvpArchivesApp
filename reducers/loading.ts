import _ from "lodash";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import * as actions from "../actions";
import { LOADING_END, LOADING_START } from "../actiontypes/loading";
import { IReducerState } from "./index";

export interface ILoadingState {
  [key: string]: boolean;
}

function loading(state: ILoadingState = {}, action: ActionType<typeof actions.loading>) {
  switch (action.type) {
    case LOADING_START:
      return {
        [action.payload]: true
      };
    case LOADING_END:
      return {
        [action.payload]: false
      };
    default:
      return state;
  }
}

export const getLoading = (loadingPrefix: string, propsPath: string, defaultValue?: any) => {
  return createSelector<IReducerState, object, ILoadingState, string, boolean>(
      [
        (state) => state.loading,
        (_state, props) => _.get(props, propsPath, defaultValue)
      ],
      // Sometimes loading will be undefined so convert to boolean
      (loadingState, loadingId) => !!loadingState[`${loadingPrefix}${loadingId}`]
  );
};

export default loading;
