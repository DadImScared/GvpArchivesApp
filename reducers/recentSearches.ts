import moment from "moment";
import { ActionType } from "typesafe-actions";

import { recentSearches } from "../actions";
import { ADD_SEARCH } from "../actiontypes/recentSearches";

export interface IRecentSearch {
  query: string;
  categories: string[];
  timestamp: number;
}

export type RecentSearches = IRecentSearch[];

export default function(state: RecentSearches = [], action: ActionType<typeof recentSearches>) {
  switch (action.type) {
    case ADD_SEARCH:
      const { payload } = action;
      return [
        {
          categories: payload.categories,
          query: payload.query,
          timestamp: moment().valueOf()
        },
        ...state
      ];
    default:
      return state;
  }
}
