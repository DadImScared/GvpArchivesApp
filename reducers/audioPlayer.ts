import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import * as actions from "../actions";
import * as AudioActionTypes from "../actiontypes/audioPlayer";
import { IReducerState } from "./index";
import { getItemsById, ItemsByIdState } from "./itemsById";

export enum ShowPlayerStatus {
  OPEN = "OPEN",
  CLOSED = "CLOSED",
  MINIMIZED = "MINIMIZED"
}

export interface IAudioPlayerState {
  currentIndex: number;
  currentSongName: string;
  currentSongUrl: string;
  items: string[];
  name: string;
  duration: number;
  position: number;
  playerType: "song" | "playlist";
  playing: boolean;
  seekTo: number;
  showPlayer: ShowPlayerStatus;
  sliderValue: number;
}

export type AudioPlayerStateAndActions<
  T extends keyof IAudioPlayerState | keyof typeof actions.audioPlayer
> = Pick<IAudioPlayerState & typeof actions.audioPlayer, T>;

export const getInitialAudioPlayerState = (): IAudioPlayerState => ({
  playerType: "song",
  playing: false,
  seekTo: 0,
  showPlayer: ShowPlayerStatus.CLOSED,
  sliderValue: 0,

  // values used when audio is playing a playlist
  currentIndex: 0,
  items: [],
  name: "",

  // values used when audio player is playing single song
  currentSongName: "",
  currentSongUrl: "",

  // playback status from expo player
  duration: 0,
  position: 0
});

const initialAudioPlayerState = getInitialAudioPlayerState();

export default function audioPlayer(
  state: IAudioPlayerState = initialAudioPlayerState,
  action: ActionType<typeof actions.audioPlayer>
) {
  switch (action.type) {
    case AudioActionTypes.UPDATE_PLAYBACKINSTANCE:
      const { payload } = action;
      if (payload.isLoaded) {
        return {
          ...state,
          duration: payload.durationMillis || 0,
          playing: payload.didJustFinish ? false : state.playing,
          position: payload.positionMillis || 0,
          sliderValue: payload.positionMillis || 0
        };
      }
      return state;
    case AudioActionTypes.SEEK_END:
      return {
        ...state,
        seekTo: action.payload.seekTo * (state.duration || 0)
      };
    case AudioActionTypes.SET_SHOW_PLAYER:
      return {
        ...state,
        showPlayer: action.payload.status
      };
    case AudioActionTypes.SEEK_TO:
      return {
        ...state,
        sliderValue: action.payload.seekTo * (state.duration || 0)
      };
    case AudioActionTypes.TOGGLE_PLAYING:
      return {
        ...state,
        playing: !state.playing
      };
    case AudioActionTypes.SET_PLAYING:
      return {
        ...state,
        playing: action.payload.playing
      };
    case AudioActionTypes.SET_PLAYLIST:
      return {
        ...state,
        currentIndex: action.payload.currentIndex,
        items: action.payload.items,
        name: action.payload.name,
        playerType: action.payload.playerType,
        playing: action.payload.playing,
        showPlayer: action.payload.showPlayer
      };
    case AudioActionTypes.UPDATE_CURRENT_INDEX:
      return {
        ...state,
        currentIndex: action.payload.newIndex
      };
    case AudioActionTypes.SET_SONG:
      return {
        ...state,
        currentSongName: action.payload.songName,
        currentSongUrl: action.payload.songUrl,
        playerType: action.payload.playerType,
        playing: action.payload.playing,
        showPlayer: state.showPlayer === ShowPlayerStatus.MINIMIZED ? state.showPlayer : action.payload.showPlayer
      };
    default:
      return state;
  }
}

const padWithZero = (time: number) => {
  if (time < 10) {
    return `0${time}`;
  }
  return time.toString();
};

const getTimestamp = (millis: number): string => {
  const totalSeconds = millis / 1000;
  const seconds = Math.floor(totalSeconds % 60);
  const minutes = Math.floor(totalSeconds / 60);

  let timeFormat = `${padWithZero(minutes)}:${padWithZero(seconds)}`;
  if (minutes > 59) {
    const hours = Math.floor(minutes / 60);
    timeFormat = `${hours}:${timeFormat}`;
  }
  return timeFormat;
};

const getAudioPlayerState = (state: IReducerState) => state.audioPlayer;

const getItems = (itemIds: string[], itemsById: ItemsByIdState) => itemIds.map((itemId) => itemsById[itemId]);

export const getButtonGroupData = createSelector(
  [getAudioPlayerState],
  ({ playing }) => ({ playing })
);

export const getTitleData = createSelector(
  [getAudioPlayerState, getItemsById],
  (audio, itemsById) => {
    const { playerType, currentSongName, currentIndex, items: itemIds } = audio;
    const items = getItems(itemIds, itemsById);
    return {
      songName: playerType === "song" ? currentSongName : items[currentIndex].title
    };
  }
);

export const getSeekerData = createSelector(
  [getAudioPlayerState],
  ({ playing, position, duration, sliderValue }) => {
    return ({
      playing,
      sliderPosition: sliderValue / duration,
      timestamp: `${getTimestamp(sliderValue)} / ${getTimestamp(duration)}`
    });
  }
);

export const getAudioPlayerData = createSelector(
  [getAudioPlayerState, getItemsById],
  (
    { playing, items: itemIds, currentSongUrl, currentIndex, playerType, seekTo },
    itemsById
  ) => {
    const items = getItems(itemIds, itemsById);
    return {
      playing,
      seekTo,
      uri: playerType === "song" ? currentSongUrl : items[currentIndex].link
    };
  }
);

export const getShowAudioPlayer = createSelector(
  [getAudioPlayerState],
  ({ showPlayer }) => ({ showPlayer })
);
