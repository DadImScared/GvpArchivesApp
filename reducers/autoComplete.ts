import moment from "moment";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import autoComplete from "../actions/autoComplete";
import * as ActionTypes from "../actiontypes/autoComplete";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "./index";
import { ILoadingState } from "./loading";
import { getQueryId, getSearch, ISearchState } from "./search";

export interface IAutoCompleteState {
  [key: string]: {
    results: string[];
    timeToLive: number;
  };
}

export const getInitialAutoCompleteState = () => ({});

export default function(state = getInitialAutoCompleteState(), action: ActionType<typeof autoComplete>) {
  switch (action.type) {
    case ActionTypes.ADD_AUTOCOMPLETE:
      const { payload: { query, categories, results } } = action;
      return {
        ...state,
        [getQueryId(query, categories)]: {
          results,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    default:
      return state;
  }
}

interface IGetAutoCompleteReturn {
  results: string[];
  timeToLive: number;
}

export interface IAutoCompleteScreenData {
  autoComplete: IGetAutoCompleteReturn;
  query: string;
  queryId: string;
  categories: string[];
  isLoading: boolean;
}

export const getAutoCompleteScreen = createSelector<
  IReducerState,
  IAutoCompleteState,
  ISearchState,
  ILoadingState,
  IAutoCompleteScreenData
>(
  [(state) => state.autoComplete, getSearch, (state) => state.loading],
  (autoCompleteState, { categories, query }, loadingState) => {
    const queryId = getQueryId(query, categories);
    return {
      categories,
      query,
      queryId,
      autoComplete: autoCompleteState[queryId] || { results: [], timeToLive: 0 },
      isLoading: loadingState[`${AUTOCOMPLETE_QUERY}${queryId}`]
    };
  }
);
