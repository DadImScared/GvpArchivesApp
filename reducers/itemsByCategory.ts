import _ from "lodash";
import moment from "moment";
import { NavigationScreenProps } from "react-navigation";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import actions from "../actions";
import { ADD_ITEMS_BY_CATEGORY, UPDATE_ITEMS_BY_CATEGORY } from "../actiontypes/itemsByCategory";
import { ITEMS_BY_CATEGORY } from "../loadingPrefix";
import { IReducerState } from "./index";
import { getItemsById, Item } from "./itemsById";
import { getLoading } from "./loading";

interface ItemsByCategory {
  results: string[];
  nextPage?: boolean | string;
  timeToLive?: number;
}

export interface ItemsByCategoryData {
  results: Item[];
  isLoading: boolean;
  nextPage: boolean | string;
  timeToLive: number;
}

export interface ItemsByCategoryState {
  [key: string]: ItemsByCategory;
}

function itemsByCategory(
    state: ItemsByCategoryState = {},
    action: ActionType<typeof actions.itemsByCategory>
) {
  switch (action.type) {
    case ADD_ITEMS_BY_CATEGORY:
      return {
        ...state,
        [action.payload.category]: {
          nextPage: action.payload.nextPage,
          results: action.payload.itemIds,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    case UPDATE_ITEMS_BY_CATEGORY:
      return {
        ...state,
        [action.payload.category]: {
          ...state[action.payload.category],
          nextPage: action.payload.nextPage,
          results: _.union(state[action.payload.category].results, action.payload.itemIds)
        }
      };
    default:
      return state;
  }
}

export const getItems = (state: IReducerState) => state.itemsByCategory;

export const getCategory = (_state: any, props: NavigationScreenProps) => props.navigation.getParam("category");

export const getItemsByCategoryState = createSelector(
    [getItems, getCategory],
    (items, category) => items[category] || {}
);

export const getItemsByCategory = () => {
  return createSelector(
      [
        getItemsByCategoryState,
        getItemsById,
        getLoading(ITEMS_BY_CATEGORY, "navigation.state.params.category")
      ],
      ({ results = [], nextPage = false, timeToLive = 0 }, items, isLoading): ItemsByCategoryData => ({
        isLoading,
        nextPage,
        timeToLive,
        results: results.map((itemId) => items[itemId])
      })
  );
};

export default itemsByCategory;
