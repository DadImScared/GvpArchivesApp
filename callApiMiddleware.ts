import {  Middleware } from "redux";
import { batchActions } from "redux-batched-actions";

import { loading } from "./actions";
import { IApiCallConfig } from "./actions/search";
import { IReducerState } from "./reducers";

const isApiCallConfig = (obj: any): obj is IApiCallConfig<any> => {
  return "apiCall" in obj;
};

export const callApiMiddleware: Middleware<{}, IReducerState> = ({ getState, dispatch }) => {
  return (next) => async (action: IApiCallConfig<any> | any) => {

    if (isApiCallConfig(action)) {
      const {
        afterApiCall = () => [],
        beforeApiCall = () => [],
        apiCall,
        loadingId = "",
        shouldCallApi = () => true
      } = action;
      const shouldLoad: string = typeof loadingId === "function" ? loadingId(getState()) : loadingId;
      // can return an axios promise to replace the original apiCall
      // used if the current data is bad
      const shouldCallOrReset = shouldCallApi(getState(), dispatch);
      let request: any;

      if (typeof shouldCallOrReset === "boolean") {
        if (shouldCallOrReset) {
          request = apiCall(getState(), dispatch);
        } else {
          return;
        }
      } else {
        request = shouldCallOrReset;
      }

      dispatch(batchActions([
        ...(await beforeApiCall(getState(), dispatch)),
        ...(shouldLoad ? [loading.loadingStart(shouldLoad)] : []),
      ]));

      const { data } = await request;

      dispatch(batchActions([
        ...(await afterApiCall(data, getState(), dispatch)),
        ...(shouldLoad ? [loading.loadingEnd(shouldLoad)] : []),
      ]));

    } else {
      return next(action);
    }
  };
};
