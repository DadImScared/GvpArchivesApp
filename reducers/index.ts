import _ from "lodash";
import { combineReducers, DeepPartial } from "redux";

import audioPlayer, { getInitialAudioPlayerState, IAudioPlayerState } from "./audioPlayer";
import autoComplete, { getInitialAutoCompleteState, IAutoCompleteState } from "./autoComplete";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";
import search, { getInitialSearchState, ISearchState } from "./search";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  autoComplete: IAutoCompleteState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
  search: ISearchState;
}

export const getInitialReducerState = (injectedState: DeepPartial<IReducerState> = {}) => _.merge({
  audioPlayer: getInitialAudioPlayerState(),
  autoComplete: getInitialAutoCompleteState(),
  itemsByCategory: {},
  itemsById: {},
  loading: {},
  search: getInitialSearchState()
}, injectedState);

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  autoComplete,
  itemsByCategory,
  itemsById,
  loading,
  search
});

export default reducer;
