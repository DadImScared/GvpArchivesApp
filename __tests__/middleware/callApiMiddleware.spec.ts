import { callApiMiddleware } from "../../callApiMiddleware";
import { getInitialReducerState, IReducerState } from "../../reducers";

const create = (middleWare = callApiMiddleware) => {
  const store = {
    dispatch: jest.fn(),
    getState: jest.fn((): IReducerState => getInitialReducerState({
      search: { query: "query" }
    }))
  };
  const next = jest.fn();

  const invoke = (action: any) => middleWare(store)(next)(action);

  return { store, next, invoke };
};

describe("callApiMiddleware", () => {
  let baseAction: any;
  let invoke: any;
  let next: any;
  let store: any;
  beforeEach(() => {
    baseAction = {
      afterApiCall: jest.fn(() => []),
      apiCall: jest.fn(() => Promise.resolve({ data: {} })),
      beforeApiCall: jest.fn(() => []),
      loadingId: "LOADING_ID_HERE"
    };
    const createMiddleware = create();
    invoke = createMiddleware.invoke;
    next = createMiddleware.next;
    store = createMiddleware.store;
  });

  it("should pass through not matching action", () => {
    const action = { type: "TEST" };
    invoke(action);
    expect(next).toHaveBeenCalledWith(action);
  });

  it("should respond to api call", async () => {
    await invoke(baseAction);
    expect(baseAction.apiCall).toHaveBeenCalledWith(store.getState(), store.dispatch);
    expect(baseAction.afterApiCall).toHaveBeenCalledWith({ data: {} }, store.getState(), store.dispatch);
    expect(store.dispatch.mock.calls).toMatchSnapshot("loading actions");
  });

  it("should call shouldCallApi function", async () => {
    const action = {
      ...baseAction,
      shouldCallApi: jest.fn(() => false)
    };
    await invoke(action);
    expect(action.shouldCallApi).toHaveBeenCalledWith(store.getState(), store.dispatch);
    expect(store.dispatch).toHaveBeenCalledTimes(0);
    expect(action.apiCall).toHaveBeenCalledTimes(0);
  });
});
