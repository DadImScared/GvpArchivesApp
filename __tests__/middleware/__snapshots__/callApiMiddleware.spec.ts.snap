// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callApiMiddleware should respond to api call: loading actions 1`] = `
Array [
  Array [
    Object {
      "meta": Object {
        "batch": true,
      },
      "payload": Array [
        Object {
          "meta": undefined,
          "payload": "LOADING_ID_HERE",
          "type": "loading/LOADING_START",
        },
      ],
      "type": "BATCHING_REDUCER.BATCH",
    },
  ],
  Array [
    Object {
      "meta": Object {
        "batch": true,
      },
      "payload": Array [
        Object {
          "meta": undefined,
          "payload": "LOADING_ID_HERE",
          "type": "loading/LOADING_END",
        },
      ],
      "type": "BATCHING_REDUCER.BATCH",
    },
  ],
]
`;
